bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections/ConcurrentLongHashMap.java

```diff
@@ -507,7 +507,11 @@ private V remove(long key, Object value, int keyHash) {
             } finally {
                 if (autoShrink && size < resizeThresholdBelow) {
                     try {
-                        int newCapacity = alignToPowerOfTwo((int) (capacity / shrinkFactor));
+                        // Shrinking must at least ensure initCapacity,
+                        // so as to avoid frequent shrinking and expansion near initCapacity,
+                        // frequent shrinking and expansion,
+                        // additionally opened arrays will consume more memory and affect GC
+                        int newCapacity = Math.max(alignToPowerOfTwo((int) (capacity / shrinkFactor)), initCapacity);
```

Comment:
Why not just skip shrink when the size is smaller than the alignToPowerOfTwo(initCapacity)? That would save a lot of computing of `alignToPowerOfTwo((int) (capacity / shrinkFactor))`

Reply:
Assuming that `initCapacity` is 256,`mapFillFactor, mapIdleFactor, expandFactor, shrinkFactor ` keep the default values:

| initCapacity | capacity | mapFillFactor | mapIdleFactor | resizeThresholdUp | resizeThresholdBelow | expandFactor | shrinkFactor |
| -----------  | -------- | ------------- | ------------- | ----------------- | -------------------- | ------------ | ------------ |
|    256       |    256   |     0.66f     |     0.15f     |         168       |           38         |      2       |        2     |
|    256       |    512   |     0.66f     |     0.15f     |         337       |           76         |      2       |        2     |
|    256       |    1024  |     0.66f     |     0.15f     |         675       |           153        |      2       |        2     |
|    256       |    2048  |     0.66f     |     0.15f     |         1351      |           307        |      2       |        2     |

When the `capacity` is expanded to 1024, the `size` needs to be less than 153 to shrink, and size=153 is less than initCapacity=256, causing at least 1024-153 of space wasted. This waste is worse when initCapacity is larger.

Therefore, we cannot use `size` to judge that it will not shrink when it is smaller than `initCapacity`.

@zymap PTAL

Reply:
This calculation will only happen if `size < resizeThresholdBelow`, it is worthwhile to use the occasional calculation to relieve the free space that is not being used.

Reply:
Thanks for your detailed explanation!